docs/proposals/20241211-reuse-nodepool-list-watch-requests.md

```diff
@@ -0,0 +1,102 @@
+
+### Goals
+1. Reduce network traffice between cloud and edge caused by workloads upgrade.
+2. Replace the old solution for resuing list/watch requests that providered by yurt-coordinator component. this means we will deprecate yurt-coordinator component.
+3. The new solution can integrate with current node-level traffic reuse solution seamlessly.
+
+### Non-Goals/Future Work
+1. The new solution can only take effect when nodes can access each other in nodepool.
+2. Don't support resuing CRD list/watch requests in this proposal. we will support this feature in the future version.
+3. Support high availability of Leader Yurthub in the future version.
+
+## Proposal
+### Metadata Types
+In the following figure, we can see that kube-proxy on every node list/watch endpointslice resource from cloud control-plane.
+
+![img.png](../img/reuse-nodepool-list-watch-requests/metadata-types.png)
+
+we can separate the metadata that components or pods on worker node list/watch from control-plane into the following two types:
+
++ Pool Scope Metadata:
+
+Each node list/watch the same copy of metadata from the control-plane, such as entire endpointslice metadata. This type of metadata is the main cause for bandwidth congestion because each change of metadata should be distributed to all nodes.
+
++ Node Scope Metadata:
+
+Each node only list/watch metadata resource related to the node, such as kubelet list/watch pods that assigned to the node. Each change of this type of metadata is only distributed the corresponding node.
+
+So we should focus on the Pool Scope Metadata. If only one copy of metadata change should be distributed to one NodePool, and nodes in NodePool share this copy of metadata, the network traffic between cloud-edge would be reduced a lot obviously.
+
+### Reuse List/Watch Requests In NodePool
+The solution is demonstrated in the following figure:
+
+![img.png](../img/reuse-nodepool-list-watch-requests/reuse-list-watch-requests.png)
+
+1. A new controller named HubLeaderElection is added in yurt-manager component. This controller will select one Yurthub as leader for each NodePool. The status of node for Leader Yurthub must be ready, if the node status becomes not ready, the controller should renew the Leader Yurthub with ready node.
+2. Different leader election strategies can be supported, such as random, mark, etc.
+    - random: controller select one ready node as leader at random.
+    - mark: controller select one ready node as leader from nodes that are specified by labelselector.
+3. Leader Yurthub will start to list/watch pool scope metadata from control-plane and store on local memory for sharing.
```

Comment:
When we refer to local memory are we talking about RAM or Disk?

Reply:
local memory here is about RAM.